airbyte-cdk/python/airbyte_cdk/sources/declarative/requesters/http_requester.py

```diff
@@ -17,64 +18,66 @@
-        :param path: Path to send requests to
-        :param http_method: HTTP method to use when sending requests
-        :param request_options_provider: request option provider defining the options to set on outgoing requests
-        :param authenticator: Authenticator defining how to authenticate to the source
-        :param error_handler: Error handler defining how to detect and handle errors
-        :param config: The user-provided configuration as specified by the source's spec
-        :param options: Additional runtime parameters to be used for string interpolation
-        """
-        if request_options_provider is None:
-            request_options_provider = InterpolatedRequestOptionsProvider(config=config)
-        elif isinstance(request_options_provider, dict):
-            request_options_provider = InterpolatedRequestOptionsProvider(config=config, **request_options_provider)
-        self._name = name
-        self._authenticator = authenticator or NoAuth()
-        self._url_base = url_base
-        self._path: InterpolatedString = path
-        if type(http_method) == str:
-            http_method = HttpMethod[http_method]
-        self._method = http_method
-        self._request_options_provider = request_options_provider
-        self._error_handler = error_handler or DefaultErrorHandler()
-        self._config = config
+    name: str
+    url_base: InterpolatedString
+    path: InterpolatedString
+    config: Config
+    options: InitVar[Mapping[str, Any]]
+    http_method: Union[str, HttpMethod] = HttpMethod.GET
+    request_options_provider: Optional[RequestOptionsProvider] = None
+    authenticator: HttpAuthenticator = None
+    error_handler: Optional[ErrorHandler] = None
+
+    def __post_init__(self, options: Mapping[str, Any]):
+        if self.request_options_provider is None:
+            self._request_options_provider = InterpolatedRequestOptionsProvider(config=self.config, options=options)
+        elif isinstance(self.request_options_provider, dict):
+            self._request_options_provider = InterpolatedRequestOptionsProvider(config=self.config, **self.request_options_provider)
+        else:
+            self._request_options_provider = self.request_options_provider
+        self.authenticator = self.authenticator or NoAuth()
+        if type(self.http_method) == str:
+            self.http_method = HttpMethod[self.http_method]
+        self._method = self.http_method
+        self.error_handler = self.error_handler or DefaultErrorHandler(options=options)
         self._options = options
 
+    # We are using an LRU cache in should_retry() method which requires all incoming arguments (including self) to be hashable.
+    # Dataclasses by default are not hashable, so we need to define __hash__(). Alternatively, we can set @dataclass(frozen=True),
+    # but this has a cascading effect where all dataclass fields must also be set to frozen.
+    def __hash__(self):
```

Comment:
This one is pretty interesting. The comment summarizes the issue, but basically, by default, dataclasses that are not marked as frozen are considered unhashable. And because an input to the `should_retry()` method is `self` our dataclass for http_requester has to hashable. Unfortunately, even if we set `frozen=True` for this class, which will now auto generate the `__hash__()` method, the various fields of the http requester are not hashable. So this has a ripple effect of forcing most underlying components to also be frozen. The problem with that, is frozen dataclasses makes all of our `__post_init__()` very unwieldly requiring everything to use `__set_attr__()`. I can go back and make everything frozen and use set_attr if we really think its critical to freeze components, but the code get's extremely ugly because some post_init functions are massive.